### PR TITLE
Fix bugs in matmul condition function and rename tests

### LINKED ISSUE #1505: Issue in gemm_to_matmul_add
Error:

```
  File "/home/xadupre/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "<eval_with_key>.4", line 5, in forward
    fused_0 = self.fused_0(primals_3, primals_1, primals_2);  primals_3 = primals_1 = primals_2 = None
  File "/home/xadupre/.local/lib/python3.10/site-packages/torch/fx/graph_module.py", line 735, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
  File "/home/xadupre/.local/lib/python3.10/site-packages/torch/onnx/_internal/onnxruntime.py", line 931, in _ort_acclerated_call
    onnx_model = optimizer.optimize(onnx_model)
  File "/home/xadupre/github/onnxscript/onnxscript/optimizer/__init__.py", line 80, in optimize
    model = rewriter.rewrite(
  File "/home/xadupre/github/onnxscript/onnxscript/rewriter/__init__.py", line 38, in rewrite
    count = pattern_rewrite_rules.apply_to_model(model_ir)
  File "/home/xadupre/github/onnxscript/onnxscript/rewriter/pattern.py", line 881, in apply_to_model
    count = self._apply_to_graph_or_function(model, model.graph)
  File "/home/xadupre/github/onnxscript/onnxscript/rewriter/pattern.py", line 871, in _apply_to_graph_or_function
    delta = rule.try_rewrite(model, graph_or_function, node)
  File "/home/xadupre/github/onnxscript/onnxscript/rewriter/pattern.py", line 743, in try_rewrite
    match = self.matches(node, model)
  File "/home/xadupre/github/onnxscript/onnxscript/rewriter/pattern.py", line 734, in matches
    and not self._condition_function(**match.bindings)
  File "/home/xadupre/github/onnxscript/onnxscript/rewriter/broadcast_to_matmul.py", line 114, in check_if_need_reshape
    broadcast_matmul_output_shape.pop(-2)
IndexError: pop index out of range
```

To reproduce:

```python
        import torch

        def _make_aot_ort(dynamic: bool = False) -> tuple:
            import onnxruntime
            from torch.onnx import (
                _OrtBackend as OrtBackend,
                _OrtBackendOptions as OrtBackendOptions,
                ExportOptions,
            )
            export_options = ExportOptions(dynamic_shapes=dynamic)
            options = OrtBackendOptions(export_options=export_options)
            ort_backend = OrtBackend(options=options)
            return ort_backend

        class Linear(torch.nn.Module):
            def __init__(self):
                super(Linear, self).__init__()
                self.linear = torch.nn.Linear(128, 10)
                self.activation = torch.nn.ReLU()

            def forward(self, *inputs):
                input = self.linear(inputs[0])
                input = self.activation(input)
                return input

        model = Linear()
        model.train()
        loss_fn = torch.nn.MSELoss()

        input = torch.randn((64, 128), requires_grad=True)
        labels = torch.randn((64, 10), requires_grad=True)

        compiled_model = torch.compile(model, backend=_make_aot_ort())
        output = compiled_model(*input)
        loss = loss_fn(output, labels)
        loss.backward()
```

### PR DESCRIPTION
**Pull request by titaiwangms**
Fix #1505 

The condition function in gemm and matmul pattern is trying to validate whether reshapes are redundant. In the previous code, we mimic the matmul behavior to get the final output shape without identifying that when input rank is 2, is it expanded with 1 from one dimension or originally is 2 dimension.

### CODE REVIEW

File: onnxscript/rewriter/broadcast_to_matmul.py [hunk 1]
```diff
@@ -108,9 +108,11 @@ def check_if_need_reshape(input_a, input_b, shape_c, **_) -> bool:
     broadcast_matmul_output_shape = (
         longer_shape[: -len(shorter_shape)] + broadcast_matmul_output_shape
     )
-    if mimic_matmul_broadcast_behavior and dim_b == 2:
+    if mimic_matmul_broadcast_behavior and dim_b == 2 and input_b_shape[-1] == 1:
+        # If input_b is expanded to 2-D, then we need to remove the last dimension
         broadcast_matmul_output_shape = broadcast_matmul_output_shape[:-1]
-    if mimic_matmul_broadcast_behavior and dim_a == 2:
+    if mimic_matmul_broadcast_behavior and dim_a == 2 and input_a_shape[0] == 1:
+        # If input_a is expanded to 2-D, then we need to remove the duplicated dimension
```
Comment (justinchuby) [inline onnxscript/rewriter/broadcast_to_matmul.py:115]: What does "duplicated dimension" mean? Was it a sized 1 dim that was broadcasted?

File: onnxscript/rewriter/gemm_to_matmul_add_test.py [hunk 2]
```diff
@@ -228,6 +228,48 @@ def test_reshape_gemm_reshape_replace_when_second_input_is_one_dimension_and_not
         self.assertEqual(count, 0)
         self.assertEqual(len(model.graph), 5)
 
+    def test_reshape_gemm_reshape_replace_when_inputs_are_two_dimension_and_broadcastable(
```
Comment (justinchuby) [inline onnxscript/rewriter/gemm_to_matmul_add_test.py:231]:
```suggestion
    def test_reshape_gemm_reshape_replaces_when_inputs_are_two_dimensional_and_broadcastable(
```

File: onnxscript/rewriter/gemm_to_matmul_add_test.py [hunk 3]
```diff
@@ -228,6 +228,48 @@ def test_reshape_gemm_reshape_replace_when_second_input_is_one_dimension_and_not
...
+                shape_a = Constant<value: tensor = int64[2] {3, 5}>()
+                reshape_x = Reshape (input_x, shape_a)
+                gemm = Gemm<alpha=1.0, beta=1.0> (reshape_x, input_y, input_z)
+                shape_d = Constant<value: tensor = int64[2] {3, 10}>()
+                output = Reshape (gemm, shape_d)
+            }
+        """
+        )
+        model = ir.serde.deserialize_model(model_proto)
+        count = gemm_to_matmul_add.rule.apply_to_model(model)
```
Comment (justinchuby) [inline onnxscript/rewriter/gemm_to_matmul_add_test.py:248]:
```suggestion
        replacement_count = gemm_to_matmul_add.rule.apply_to_model(model)
```

for clarity?

**Review by justinchuby — APPROVED** [review]

File: onnxscript/rewriter/broadcast_to_matmul.py (hunk 1 above)
Comment (titaiwangms) [reply onnxscript/rewriter/broadcast_to_matmul.py]: Changed it to the first dim of input_a, which is -2nd dim of the mimic output shape.

**Pull request merged**